src/main/java/com/limechain/network/protocol/grandpa/GrandpaEngine.java

```diff
@@ -0,0 +1,104 @@
+package com.limechain.network.protocol.grandpa;
+
+import com.limechain.network.ConnectionManager;
+import com.limechain.network.PeerInfo;
+import com.limechain.network.protocol.grandpa.messages.GrandpaMessageType;
+import com.limechain.network.protocol.grandpa.messages.commit.CommitMessage;
+import com.limechain.network.protocol.grandpa.messages.commit.CommitMessageScaleReader;
+import com.limechain.network.protocol.grandpa.messages.neighbour.NeighbourMessage;
+import com.limechain.network.protocol.grandpa.messages.neighbour.NeighbourMessageScaleReader;
+import com.limechain.network.protocol.grandpa.messages.neighbour.NeighbourMessageScaleWriter;
+import com.limechain.sync.warpsync.SyncedState;
+import io.emeraldpay.polkaj.scale.ScaleCodecReader;
+import io.emeraldpay.polkaj.scale.ScaleCodecWriter;
+import io.libp2p.core.PeerId;
+import io.libp2p.core.Stream;
+import lombok.extern.java.Log;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.logging.Level;
+
+@Log
+public class GrandpaEngine {
+    private static final int HANDSHAKE_LENGTH = 1;
+    private final ConnectionManager connectionManager = ConnectionManager.getInstance();
+    private final SyncedState syncedState = SyncedState.getInstance();
+
+    public void receiveRequest(byte[] msg, PeerId peerId, Stream stream) {
+        boolean connectedToPeer = connectionManager.isGrandpaConnected(peerId);
+        GrandpaMessageType messageType = getGrandpaMessageType(msg);
+
+        if (!connectedToPeer && messageType != GrandpaMessageType.HANDSHAKE) {
+            log.log(Level.WARNING, "No handshake for grandpa message from Peer " + peerId);
+            return;
+        }
+
+        switch (messageType) {
+            case HANDSHAKE -> handleHandshake(msg, peerId, stream);
+            case NEIGHBOUR -> handleNeighbourMessage(msg, peerId, stream);
+            case COMMIT -> handleCommitMessage(msg, peerId);
+            default -> log.log(Level.WARNING, "Unknown grandpa message type from Peer " + peerId);
+        }
+    }
```

Comment:
I think we should add the other message types as well and only log that they've been received